docroot/modules/custom/mass_entity_usage/mass_entity_usage.module

```diff
@@ -0,0 +1,116 @@
+use Drupal\mayflower\Helper;
+
+/**
+ * Implements hook_entity_usage_block_tracking().
+ */
+function mass_entity_usage_entity_usage_block_tracking($target_id, $target_type, $source_id, $source_type, $source_langcode, $source_vid, $method, $field_name, $count) {
+  // Block tracking of entities with a node source that is unpublished or trash.
+  if (!_mass_entity_usage_tracking_check($source_type, $source_id, $source_vid)) {
+    return TRUE;
+  }
+  return FALSE;
+}
+
+/**
+ * Find an id/revision_id record in the default field data table.
+ *
+ * @param string $entity_type
+ *   The entity type.
+ * @param int $id
+ *   The entity id.
+ * @param int $vid
+ *   The entity revision id.
+ *
+ * @return bool
+ *   If true, the content should be tracked. If false, do not track.
+ */
+function _mass_entity_usage_tracking_check($entity_type, $id, $vid) {
+  if ($entity_type == 'paragraph') {
+    // Verify the paragraph revision is in the paragraphs_item_field_data table.
+    $query = \Drupal::database()->select('paragraphs_item_field_data', 'pifd');
+    $query->fields('pifd', ['id', 'revision_id']);
+    $query->condition('id', $id);
+    $query->condition('revision_id', $vid);
+    $results = $query->execute()->fetchAll();
+    if (empty($results)) {
+      return FALSE;
+    }
+    // Load the paragraph revision.
+    $paragraph = \Drupal::service('entity_type.manager')->getStorage('paragraph')
+      ->loadByProperties(['id' => $id, 'revision_id' => $vid]);
+    // Get the parent node of the paragraph.
+    $parent_node = Helper::getParentNode(current($paragraph));
+    if (is_null($parent_node)) {
+      return FALSE;
+    }
+    // Check if the parent node should be tracked.
+    return _mass_entity_usage_tracking_check('node', $parent_node->id(), $parent_node->getRevisionId());
+  }
+  if ($entity_type == 'node') {
+    // Load the node.
+    $node = \Drupal::entityTypeManager()->getStorage('node')->loadRevision($vid);
+    // Get the moderation state.
+    $content_moderation_state = ContentModerationState::loadFromModeratedEntity($node);
+    if ($content_moderation_state instanceof ContentModerationStateInterface
+      && !$content_moderation_state->get('moderation_state')->isEmpty()) {
+      $state_name = $content_moderation_state->get('moderation_state')->value;
+      // These states should be not be tracked.
+      $skipped_states = [
+        MassModeration::UNPUBLISHED,
+        MassModeration::PREPUBLISHED_DRAFT,
+        MassModeration::DRAFT,
+        MassModeration::PREPUBLISHED_NEEDS_REVIEW,
+        MassModeration::TRASH,
+      ];
+      // Don't track sources that are in unpublished or trash state.
+      if (in_array($state_name, $skipped_states)) {
+        return FALSE;
+      }
+    }
+  }
+  return TRUE;
+}
+
+/**
+ * Implements hook_entity_insert().
+ */
+function mass_entity_usage_entity_insert(EntityInterface $entity) {
+  Drupal::queue('mass_entity_usage_delete_not_current_revision_usages')->createItem([
```

Comment:
What queues does this feature use and why does each exist? 

For example, I cant think of a reason why regenerate (i.e. the bulk recalculation) has to be a different queue from the entity add/edit queue.

Also, I cant know why mass_entity_usage_delete_not_current_revision_usages is queued in the first place. Didn't we develop a function which deletes all non-current revisions for any node? Maybe I'm misremembering that.

Reply:
I think you are referring to mass_hierachy mass_hierarchy_delete_all_non_current_revisions_from_nested_set_table, that executes on mass_hierarchy_cron. That is not about usages, but hierarchy.

Regarding this queue `mass_entity_usage_delete_not_current_revision_usages`, is to delete usages of non-published revisions for nodes as we don't need that information and it was showing misleading references in our entity usage reports.

Reply:
Thanks. Could we have a query that runs once a day and removes non-current revisions? If it cant delete immediately, it could queue up the deletes. The improvement is that we dont do any enqueue on each entity CRUD.

Reply:
I think we are not processing queues on the entities' CRUD. We are just adding items to the queue. And because we are using queues we have to manually add a scheduled task to do this, which gives us control over its execution time and frequency. @weitzman

Reply:
Yes, that makes sense. But my suggestion is more robust. Consider what happens if we have to empty the deletion queue for some reason. With the current solution, will have keep non-current revisions in the table for a very long time (until each one of them is next edited). With my suggestion, it is no problem. We will re-enqueue the non-current revisions immediately.

Reply:
Oh! I get it. Makes sense.